packages/tpetra/core/src/Tpetra_Distributor.hpp

```diff
@@ -2723,6 +3074,112 @@ namespace Tpetra {
+            for (size_t j = 0; j < lengthsFrom_[i]; ++j) {
+              totalPacketsFrom_i += numImportPacketsPerLID[curLIDoffset+j];
+            }
+            curLIDoffset += lengthsFrom_[i];
+            if (procsFrom_[i] != myProcID && totalPacketsFrom_i) {
+              imports_view_type recvBuf =
+                subview_offset (imports, curBufferOffset, totalPacketsFrom_i);
+
+              Packet t;
+              MPI_Datatype rawType = Teuchos::Details::MpiTypeTraits<Packet>::getType (t);
+              Packet* rawRecvBuf = recvBuf.data ();
+              const int count = as<int> (recvBuf.size ());
+              MPI_Request rawRequest = MPI_REQUEST_NULL;
+              MPI_Recv_init (rawRecvBuf, count, rawType, procsFrom_[i], tag, rawComm, &rawRequest);
+
+              Teuchos::RCP<Teuchos::CommRequest<int> > req = Teuchos::mpiCommRequest<int>(rawRequest, count);
+              // Teuchos::set_extra_data( globalRecvBuf, "buffer", inOutArg(req) );
+              persistentRequests_.push_back(req);
+            }
+            else { // Receiving these packet(s) from myself
+              selfReceiveOffset = curBufferOffset; // Remember the offset
+            }
+            curBufferOffset += totalPacketsFrom_i;
+          }
+
+          if (verbose_) {
+            std::ostringstream os;
+            os << "Proc " << myProcID
+               << ": doPosts(4 args, Kokkos, fast): Creating persistent send requests" << endl;
+            *out_ << os.str ();
+          }
+          // create the persistent send requests
+          for (size_t i = 0; i < numBlocks; ++i) {
+            size_t p = i + procIndex;
+            if (p > (numBlocks - 1)) {
+              p -= numBlocks;
+            }
+
+            if (procsTo_[p] != myProcID && packetsPerSend[p] > 0) {
+
+              exports_view_type sendBuf =
+                subview_offset (exports, sendPacketOffsets[p], packetsPerSend[p]);
+
+              Packet t;
+              MPI_Datatype rawType = Teuchos::Details::MpiTypeTraits<Packet>::getType (t);
+              const Packet* rawSendBuf = sendBuf.data ();
+              const int count = as<int> (sendBuf.size ());
+              MPI_Request rawRequest = MPI_REQUEST_NULL;
+              MPI_Send_init (rawSendBuf, count, rawType, procsTo_[p], tag, rawComm, &rawRequest);
+
+              Teuchos::RCP<Teuchos::CommRequest<int> > req = Teuchos::mpiCommRequest<int>(rawRequest, count);
+              // Teuchos::set_extra_data( sendBuf, "buffer", inOutArg(req) );
+              persistentRequests_.push_back(req);
+            }
+          }
+        }
+
+        if (verbose_) {
+          std::ostringstream os;
+          os << "Proc " << myProcID
+             << ": doPosts(4 args, Kokkos, fast): Starting persistent requests" << endl;
+          *out_ << os.str ();
+        }
+        int count = persistentRequests_.size();
+        Array<MPI_Request> rawMpiRequests (count, MPI_REQUEST_NULL);
```

Comment:
See notes above.  We have a `Teuchos::waitAll`; a `Teuchos::startAll` would be the next logical step.